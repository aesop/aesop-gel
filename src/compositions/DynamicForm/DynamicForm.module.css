@import 'styles/index';

.base {
  display: flex;
  min-width: 400px;
  flex-direction: column;

  > div {
    margin: 10px 0;
  }

  > button {
    align-self: center;
    margin: 10px 0;
  }
}
