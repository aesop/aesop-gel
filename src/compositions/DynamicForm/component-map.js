/* eslint-disable react/prop-types */
import React from 'react';
import { Controller } from 'react-hook-form';

import { Checkbox } from '~/components/Checkbox';
import { Select } from '~/components/Select';
import { TextInputV2 } from '~/components/TextInputV2';

export const componentMap = {
  Checkbox: ({
    control,
    defaultValue,
    errorMessage,
    id,
    label,
    name,
    rules,
    theme,
  }) => (
    <Controller
      control={control}
      defaultValue={defaultValue || false}
      name={name}
      render={({ onChange }) => (
        <Checkbox
          content={label}
          errorMessage={errorMessage}
          id={id}
          onChange={e => onChange(e.target.checked)}
          theme={theme}
        />
      )}
      rules={rules}
    />
  ),
  Select: ({
    control,
    defaultValue,
    errorMessage,
    id,
    label,
    name,
    options,
    rules,
    theme,
  }) => (
    <Controller
      control={control}
      defaultValue={defaultValue || ''}
      name={name}
      render={({ onChange, value }) => (
        <Select
          errorMessage={errorMessage}
          id={id}
          label={label}
          name={name}
          onChange={onChange}
          options={options}
          theme={theme}
          value={value}
        />
      )}
      rules={rules}
    />
  ),
  TextField: ({
    control,
    defaultValue,
    errorMessage,
    id,
    label,
    name,
    rules,
    subtype: type,
    theme,
  }) => (
    <Controller
      control={control}
      defaultValue={defaultValue || ''}
      name={name}
      render={({ onChange, value }) => (
        <TextInputV2
          errorMessage={errorMessage}
          id={id}
          label={label}
          onChange={onChange}
          theme={theme}
          type={type}
          value={value}
        />
      )}
      rules={rules}
    />
  ),
};
