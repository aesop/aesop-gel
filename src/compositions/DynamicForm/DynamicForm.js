import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useForm } from 'react-hook-form';
import { useThemeContext } from '~/contexts';
import { componentMap } from './component-map';
import styles from './DynamicForm.module.css';

const DynamicForm = forwardRef(function DynamicFormRef(
  { children, className, formSchema, onSubmit, theme },
  ref,
) {
  const { control, handleSubmit, errors } = useForm({
    mode: 'onChange',
  });
  const currentTheme = useThemeContext(theme, 'dark');

  if (!formSchema) {
    return null;
  }

  const classSet = cx(styles.base, styles[currentTheme], className);

  return (
    <form className={classSet} onSubmit={handleSubmit(onSubmit)} ref={ref}>
      {formSchema.map(
        (
          { name, id, label, validation, type, options, subtype, defaultValue },
          index,
        ) => {
          const InputField = componentMap[type];

          if (!InputField) {
            // eslint-disable-next-line no-console
            console.warn(`Field with type ${type} has no registered component`);
            return null;
          }

          const rules = {};

          if (validation.isRequired?.value) {
            rules.required = validation.isRequired.message;
          }

          if (validation.pattern) {
            rules.pattern = {
              value: new RegExp(validation.pattern.value),
              message: validation.pattern.message,
            };
          }

          rules.maxLength = validation.maxLength;

          return (
            <InputField
              control={control}
              defaultValue={defaultValue}
              errorMessage={errors[name]?.message}
              id={id}
              key={index}
              label={label}
              name={name}
              options={options}
              rules={rules}
              subtype={subtype}
              theme={currentTheme}
            />
          );
        },
      )}
      {children}
    </form>
  );
});

DynamicForm.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  formSchema: PropTypes.array,
  onSubmit: PropTypes.func.isRequired,
  theme: PropTypes.oneOf(['dark', 'light']),
};

DynamicForm.defaultProps = {
  children: undefined,
  className: undefined,
  formSchema: undefined,
  onSubmit: undefined,
  theme: undefined,
};

export { DynamicForm };
