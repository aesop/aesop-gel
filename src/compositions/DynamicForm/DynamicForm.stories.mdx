import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { DynamicForm } from './DynamicForm';
import { DynamicFormFixture } from './DynamicForm.fixture';
import { Button } from '../../components/Button';

<Meta title="Compositions/DynamicForm" component={DynamicForm} />

# DynamicForm

<Canvas>
  <Story name="Base component">
    <DynamicForm
      defaultValues={{ title: 'mr', firstname: 'Bruce' }}
      formSchema={knobs.select('formSchema', {
        'None': null,
        'Au Sign up form': DynamicFormFixture.auSignUp,
        'Korean Sign up form': DynamicFormFixture.krSignUp,
      }, null, 'Content')}
      onSubmit={DynamicFormFixture.onSubmit}
      theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
    >
      <Button theme="dark" type="submit">
        Submit
      </Button>
    </DynamicForm>
  </Story>
</Canvas>

## API

<ArgsTable of={DynamicForm} />
