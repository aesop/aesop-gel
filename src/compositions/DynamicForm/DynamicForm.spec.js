import React from 'react';
import { render } from '@testing-library/react';
import { DynamicForm } from './DynamicForm';

describe('<DynamicForm />', () => {
  it('should be defined', () => {
    expect(DynamicForm).toBeDefined();
  });

  it('should render base component correctly', () => {
    const { container } = render(<DynamicForm />);

    expect(container).toMatchSnapshot();
  });
});
