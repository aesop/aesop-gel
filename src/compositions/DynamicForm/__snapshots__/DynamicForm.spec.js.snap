// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicForm /> should render base component correctly 1`] = `<div />`;
